src/main/java/org/spongepowered/api/plugin/Plugin.java

```diff
@@ -24,19 +24,21 @@
  */
 package org.spongepowered.api.plugin;
 
-import static java.lang.annotation.ElementType.TYPE;
-import static java.lang.annotation.RetentionPolicy.RUNTIME;
-
+import java.lang.annotation.ElementType;
 import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
 import java.lang.annotation.Target;
+import java.util.regex.Pattern;
 
 /**
  * An annotation used to describe and mark a Sponge plugin.
  */
-@Target(TYPE)
-@Retention(RUNTIME)
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.TYPE)
 public @interface Plugin {
 
+    Pattern ID_PATTERN = Pattern.compile("[a-z0-9-_]+");
```

Comment:
There is a **lot** of line noise in that issue and most of it is not productive in the slightest. Like it or not a mod id has to exist for implementation reasons and whether it's defined in multiple parts or not it **must** resolve to a single string in the end because that string is needed by parts of the underlying implementation.

> I'm not against allowing qualified plugin IDs, but there were people against this in #1083

I don't see anyone against qualified ids in that issue, and whilst there is a lot of unrelated noise and unnecessary prevarication about utterly unrelated behavioural decisions, the actual question of what to use as an id basically got buried. 

>  so I'd personally like to finish discussing the need for additional identifiers of plugin IDs before making any changes

But changes are exactly what you're making here. viz. currently there is no restriction on id, and you are proposing to add some.

A fully-qualified string fits all the criteria of being suitably unique, working as an id in the underlying implementation, and being familiar to existing users. In my opinion we should support this as a bare **minimum**. Our recommendations should be that plugin authors use an FQ name or a generated GUID as their ID, and then just move on with our collective lives.

The *class-as-an-identifier* thing proposed in #1083 only works if we convert it to a string internally anyway, and so the entire proposal is moot. The discussions in that thread just get more navel-gazey the more you read.

Can we just put this to bed once and for all.